### PR TITLE
Add oauth2 password

### PR DESCRIPTION
**Pull request by BirdHighway**
**What has changed?**

- Addition of new auth client type for OAuth2 Password flows
- Additon of new secret type for OAuth2 Password flows
- Updated OpenAPI docs for the new auth client and secret types

**Details**

`POST` request to `/auth-clients` endpoint to generate OAuth2 Password auth client
```json
{
    "name": "OA2 Password Test Client, creds in ac",
    "type": "OA2_PASSWORD",
    "clientId": "abc123",
    "clientSecret": "def456",
    "endpoints": {
      "token": "https://company1.sandbox.my.salesforce.com/services/oauth2/token"
    },
    "includeCredentialsInHeader": false,
    "username": "user123",  // optional
    "password": "uncrackable2" // optional
  }
```

Response:
```json
{
    "data": {
        "clientId": "***", // data is masked in response and in context of listing all secrets
        "clientSecret": "***",
        "endpoints": {
            "token": "https://company1.sandbox.my.salesforce.com/services/oauth2/token"
        },
        "includeCredentialsInHeader": false,
        "username": "idinteractive.user@harman.com.uat",
        "password": "***",
        "_id": "66919418503b5dd585f7c802",
        "name": "OA2 Password Test Client",
        "owners": [
            {
                "id": "66859956557cfc511d4000aa",
                "type": "USER"
            }
        ],
        "type": "OA2_PASSWORD",
        "__t": "A_OA2_PASSWORD",
        "createdAt": "2024-07-12T20:37:44.511Z",
        "updatedAt": "2024-07-12T20:37:44.511Z",
        "__v": 0
    }
}
```

`POST` request to `/secrets` endpoint to generate a secret:

```json
{
  "authClientId": "66919418503b5dd585f7c802",
  "name": "OA2 Pass Secret Test",
  "type": "OA2_PASSWORD",
  "username": "user123", // if not provided to auth client, or if override is desired
  "password": "uncrackable2", // if not provided to auth client, or if override is desired
}
```

Upon receipt of the secret generation request, the Secret Service will use the combined info from the auth client and the secret to make an OAuth2 Password request to the token endpoint specified in the auth client. The acccess token will be encrypted and stored in the secret.

Response - note that the JSON stringified full response from the access token endpoint is saved. This allows flows to make use of non-standard fields in the response if necessary. When the secret it fetched individually the encrypted fields will be decrypted (following the same pattern as other auth client flows):
```json
{
    "data": {
        "value": {
            "authClientId": "66919418503b5dd585f7c802",
            "accessToken":"bÉ©«!èF?G³§]:¤8þaf'¢é$¿Mu\bÂñåsxØuô2Ü½ñ\f/5øÈj-êºl%ÝÏö a°(ð¨1®\u000br^Ól« `RíC§Í»ù»ñ>Ø|ät´®LR¤2ºP×D\f",
            "tokenType": "Bearer",
            "expires": "33658-09-27T01:46:40+00:00",
            "fullResponse":"CnZjêõ©Q9JÙIq¢òm¯$\u000bãÃeFòO§ü¾G^Þùe¨¥h¯BÏ¾É\f×Å³vì^K7Q÷iJ,T%\rÇQ¡XA[° $]_ýÞAê0gí oîè>øburäÊUp*wQ4÷Ç$¿ÔçM}\sà}«ZUÍ^Ïóé^8Ï_>Ø¯Çoo¬ügÀ¬°iTÒcLCÓI=&ê;h(2[í\\±$ù/lºqRs7âÍVz}.Qu#wíh_¶e+1R¶P Å\u000bW÷þ\ts¬ç»ÓÖú2g\r'P¬ý¨þsÇgÎõ,°Î\fám©ò9D¥øéjµñ¥Î;C¤S#w.T{AÎ,¢Ù|W0"
        },
        "_id": "6691942f503b5dd585f7c805",
        "name": "OA2 Pass Secret Test",
        "owners": [
            {
                "id": "66859956557cfc511d4150db",
                "type": "USER"
            }
        ],
        "type": "OA2_PASSWORD",
        "encryptedFields": [
            "accessToken",
            "fullResponse"
        ],
        "__t": "S_OA2_PASSWORD",
        "createdAt": "2024-07-12T20:38:07.817Z",
        "updatedAt": "2024-07-12T20:38:07.817Z",
        "__v": 0
    }
}
```

**Does a specific change require especially careful review?**

No

**What is still open or a known issue?**

The OAuth2 Password flow does not typically have a refresh token (tokens do not expire), so refresh functionality is not currently enabled.

**Release Notes**

Adds support for OAuth2 Password authentication

### CODE REVIEW

File: services/secret-service/src/constant/index.js [hunk 1]
```diff
@@ -47,6 +47,7 @@ module.exports = {
             'username',
             'payload',
             'inputFields',
+            'fullResponse',
```
Comment (winklerj) [inline services/secret-service/src/constant/index.js:50]: Does `clientSecret` and `password` need added to this list?

File: services/secret-service/src/route/auth-clients/index.js [hunk 2]
```diff
@@ -26,6 +27,13 @@ const authClientObfuscator = {
         clientId: '***',
         clientSecret: '***',
     }),
+
+    [OA2_PASSWORD]: (value) => ({
+        ...value,
+        clientId: '***',
+        clientSecret: '***',
+        password: '***',
```
Comment (winklerj) [inline services/secret-service/src/route/auth-clients/index.js:35]: Does fullResponse need added to this?

File: services/secret-service/src/route/auth-clients/index.js [hunk 3]
```diff
@@ -257,6 +265,45 @@ class AuthClientRouter {
                 });
             }
         });
+
+        this.router.post('/:id/get-token', getKeyParameter, jsonParser, async (req, res, next) => {
```
Comment (winklerj) [inline services/secret-service/src/route/auth-clients/index.js:269]: Is this still needed? Would it just be the same as retrieving the secret?

File: services/secret-service/src/route/callback/handle-oauth2-password.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    const authClient = await AuthClientDAO.findById(authClientId);
+
+    if (!authClient) {
+        throw new Error('AuthClient not found');
+    }
+
+    const resolvedUsername = username || authClient.username;
+    const resolvedPassword = password || authClient.password;
+    const resolvedScope = scope || authClient.scope;
```
Comment (winklerj) [inline services/secret-service/src/route/callback/handle-oauth2-password.js:18]: I may have missed it, but isn't `authClient.scope` actually `authClient.predefinedScope`?

**Review by winklerj — CHANGES_REQUESTED** [review]

File: services/secret-service/src/route/auth-clients/index.js (hunk 3 above)
Comment (BirdHighway) [reply services/secret-service/src/route/auth-clients/index.js:269]: Good catch

File: services/secret-service/src/route/callback/handle-oauth2-password.js (hunk 4 above)
Comment (BirdHighway) [reply services/secret-service/src/route/callback/handle-oauth2-password.js:18]: That's right, thanks for catching

File: services/secret-service/src/route/auth-clients/index.js (hunk 2 above)
Comment (BirdHighway) [reply services/secret-service/src/route/auth-clients/index.js:35]: `fullResponse` is only on the secret, not on the auth client, and I believe these values only apply to auth clients.

**Review by winklerj — APPROVED** [review]

**Comment by BirdHighway** [comment]
Adding a note that the token for the secret can be force refetched by sending a `PATCH` request to `/secrets/:id` with a body specifying the type as `OA2_PASSWORD` and the `authClientId`, like this:

```json
{
    "type": "OA2_PASSWORD",
    "authClientId": "671700756eeaf82fd5109f67"
}
```

**Pull request merged**